File: leetcode/Algorithms/416.PartitionEqualSubsetSum/Solution.java
public class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int n : nums) sum += n;
        return (sum & 1) == 0 && helper(nums, 0, (sum >> 1));
    }

    private boolean helper(int[] nums, int begin, int target) {
        if (target <= 0 || begin >= nums.length) return target == 0;
        return helper(nums, begin + 1, target - nums[begin]) || helper(nums, begin + 1, target);
    }
}

File: leetcode/Algorithms/416.PartitionEqualSubsetSum/Solution.py
class Solution(object):
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        count = sum(nums)
        if count & 1 == 1:
            return False
 
        half = count >> 1
        dp = [False] * (half + 1)
        dp[0] = True
 
        for n in nums:
            for i in range(half - n, -1, -1):
                dp[i + n] |= dp[i]
 
        return dp[half]

